### ISSUE TITLE
Support and/or exploit ipv6

### ISSUE THREAD

**Issue by bgrant0607** [kubernetes/kubernetes#1443]
Self explanatory. @MalteJ mentioned ipv6 in #188, and multiple partners have mentioned ipv6 as an attractive solution for the k8s networking model, which allocates ip addresses fairly freely, for both pods and (with ip-per-service) services.

**Comment by MalteJ** [comment]
OK, the Docker IPv6 pull request is merged. Now it's time for k8s IPv6 ;)

**Comment by aanm** [comment]
Any ETA?

**Comment by MalteJ** [comment]
I don't think anyone is working on this.

**Comment by pires** [comment]
GCE and AWS don't support IPv6 so I'd figure this is very-low priority stuff.

**Comment by SuperQ** [comment]
What part of IPv6 isn't working?  I've been able to scrape targets over v6
juts fine.

For example:
  target_groups:
  - targets: ['[::1]:9090']

There are some strangeness with host:port specs in Go.
https://golang.org/src/net/ipsock.go#L107

**Comment by SuperQ** [comment]
I'm sorry, too many mailing lists.  I got this confused with the Prometheus
list. *SIGH* :)

**Comment by lattwood** [comment]
Looks like IPv6 is available in Ohio now.

https://aws.amazon.com/blogs/aws/new-ipv6-support-for-ec2-instances-in-virtual-private-clouds/

**Comment by philips** [comment]
From @thockin [via twitter](https://twitter.com/thockin/status/804854680945770496) on known things that would have to happen to make this work

- API: IP & CIDR fields
- iptables kubelet & proxy
- CNI & bridge driver
- grep and fix all places we have To4, ParseIP, ParseCIDR

**Comment by thockin** [comment]
To be clear: those are the STARTING points for auditing v6 support.  I
would approach it by making pods work, then making services work, then
finding stragglers.

**Comment by pdecat** [comment]
Now available in 15 AWS regions:

https://aws.amazon.com/blogs/aws/aws-ipv6-update-global-support-spanning-15-regions-multiple-aws-services/

**Comment by leblancd** [comment]
Before we start auditing and/or implementing the various IPv6 pieces outlined above, I think we need to come to an agreement on the following: Do we impose a restriction on the maximum IPv6 prefix length (i.e. minimum IPv6 subnet space) that is allocated to a node? My strong preference would be to limit the length of an IP prefix that is allocated to a node to be 64 bits of prefix or shorter (i.e. subnet space equivalent to a /64 space or larger). This would allow for 64 bits of interface ID, which is the defacto standard for IPv6. I believe that this is a reasonable restriction, but I haven't seen it stated/written anywhere in Kubernetes or CNI documentation. There are several reasons for imposing this restriction:
- Reasons outlined in RFC7421. Especially considering IPv6 features listed in Section 4.1 that depend on there being a 64-bit interface ID. Not sure if we'd need any particular feature listed here, but the code would be more future-proof if it allows for a 64-bit interface ID.
- To avoid collisions in the IPv6 Neighbor Discovery cache: See the discussion on this CNI pull request: https://github.com/containernetworking/cni/pull/394.
- (Less important) To avoid crazy 128-bit arithmetic operations when selecting an IPv6 subnet for a node. For example, there's this line in the function newCIDRset() function in pkg/controller/node/cidr_set.go:
    maxCIDRs := 1 << unit32(subNetMaskSize - clusterMaskSize)
This 32-bit operation currently works only for IPv4. For IPv6, this would have to be a 128-bit operation assuming that there was no restriction on subNetMaskSize (and Go language doesn't have a native 128-bit type). If subNetMaskSize is restricted to a max of 64, then a uint64 would work in place of the uint32.

Allocating a /64 space per node may sound wasteful, but it's very reasonable when you're starting with a /48 or a /50 cluster space. A /48 or /50 cluster space is reasonable esp. when it's private (a.k.a. ULA) address space.

Aside from this /64-vs-no-limit issue, some colleagues and I have started looking into the instances of To4(), ParseIP(), and ParseCIDR() in the kubernetes code, and how these might have to change for IPv6/dual-stack support. The not-so-good news is that there are lots of instances of where these are called. The good news is that 'net' library calls such as ParseIP() and ParseCIDR() are indifferent to whether they're operating on IPv6 or IPv4 addresses... they will do the right thing according to what they're passed. Similarly, the 'net' structures such as IP and IPNet work equally well for IPv4 and IPv6. Another help is that IPv4 addresses can be represented internally by their IPv4-mapped IPv6 address ::ffff:<ipv4-address> (see RFC4291, Sect. 2.5.5.2) using a 16-byte slice. In this way, a 'net' IP address can hold either:
- 4-byte slice, IPv4
- 16-byte slice, IPv4-mapped IPv6
- 16-byte slice, IPv6 address
The 'net' functions/utilities can work with any of these... it can differentiate between these as needed. In fact, the To4() function can be used to distinguish between the last 2 bullets above.
For existing calls to To4(), these can probably be replaced with a call to To16(), or just remove the call to IP4() and leave the IP unmodified.

**Comment by pires** [comment]
> Do we impose a restriction on the maximum IPv6 prefix length (i.e. minimum IPv6 subnet space) that is allocated to a node?

I believe we should.

> My strong preference would be to limit the length of an IP prefix that is allocated to a node to be 64 bits of prefix or shorter (...)

You've outlined more details than I could think of. The arithmetic concerns are spot on!

Now, from the top of my head, I think most issues will happen with:
* kube-proxy and the kubelet, which expose a few flags related to IP addresses + deal a lot with IPTables - I'm left wondering how to properly manage virtual IPv6 addresses (for services);
* CNI plug-ins;
* Bootstrap new clusters & upgrade story (if we'll be supporting it) for existing IPv4 clusters:
  * how addresses are managed, i.e. persisted in storage;
  * move components from IPv4 to IPv6, e.g. `controller-manager --cluster-cidr` where this component stops managing IPv4 and now manages IPv6
  * A potential dependency on external DNS, since:
    * component configuration still depends on knowing IPs beforehand, e.g. `kubelet --cluster-dns=10.100.0.10`
    * or `kubeadm join --discovery-token 123456.abcdefghij <one or more apiserver IPv6 addresses>`

**Comment by valentin2105** [comment]
Hi !
I'm really interested about IPv6 in k8s bare-metal cluster, 

With 1.7.x, I'm able to reach APIserver via IPv6 and to reach kubelet in IPv6 (from managers).

I'm blocked with : 

- Expose nodeport in IPv6 (it bind on, but no answer while reach them) 
- Got IPv6 in pods network. (using Calico). the most important part. 

Any people got solutions for the last two points ?

**Comment by hagbard5235** [comment]
valentin2105: I'm curious why you want node port with v6?  

Ed

**Comment by anguslees** [comment]
@hagbard5235: I'm curious why you would *not* want a node port with v6 :P 
Eventually (theoretically) _all_ IP networking will be IPv6 - so I feel like "why does _this_ specific piece need v6?" doesn't need to be justified any further.

More concretely, _I_ would like to have IPv6 nodeports, just because it's easier to manage an IPv6-only network at the host-level (more private addresses to work with, and better network bootstrap protocols).  This necessarily requires IPv6 nodeports so I can continue to use loadbalancers over such a network.

**Comment by mithrandi** [comment]
It seems like you should be able to use ClusterIP services instead of NodePort services. The problem with ClusterIP services on IPv4 is that they are not routable from outside the k8s cluster, but for IPv6 it seems like you could give everything publically routable IPv6 addresses, thus avoiding this problem.

**Comment by hagbard5235** [comment]
@anguslees Part of the reason I asked is that we have a whole bunch of practices that have evolved over time to compensate for limitations in IPv4 that are not present in IPv6.  For a hysterical example, see [this xtranormal video on IPv6 NATs](https://www.youtube.com/watch?v=v26BAlfWBm8).  On the flip side... I've found it dangerous to simply imagine things to fall into this category, because all to often there's something not in my imaginings :)  So I ask :)

Back to your concrete example, could you on your desire to manage 'at host-level'?  And why would you expect to be using IPv6 private addresses?

**Comment by hagbard5235** [comment]
@mithrandi I was kind of thinking the same thing re using ClusterIP services instead of NodePort... but I asked because (as noted above) I'd like to get a view from others on that point.  I tend to (perhaps naively) think you could simply draw your service IPs from a ServicePrefix::, Anycast that ServicePrefix:: to all Nodes, and get basically the same result as NodePorts (while letting your Network do your Node level LBing for you).  Obviously, you would still need the Nodes to "do the right thing" if the service were not available on the node receiving the traffic.

**Comment by leblancd** [comment]
@mithrandi @anguslees I have the IPv6 NodePorts (and externalIPs) working in the lab, and I'll be pushing up a PR in the next day or two (works in conjunction with PR #48551). This would allow you to configure your cluster similar to what we have for IPv4 services if you want. (There's nothing preventing you from using global cluster/pod IPs and routing directly to those, if you prefer.) For those who want to go with the IPv4 model, the address assignments would be:
    Pod IPs: ULA (or use global, if you prefer)
    Node IPs: GUA
    Service IP range: ULA
    ExternalIPs: GUA
This would allow kube-proxy to do load balancing to services via IPv6, it would keep your pod/cluster IP space private, and you would only expose the services you want exposed via nodePort or externalIPs.

**Comment by anguslees** [comment]
@hagbard5235 (and anyone else) we shouldn't use a github issue as a chat forum - so let's continue discussions on slack or elsewhere (I'm @guslees on k8s slack fwiw).

**Comment by valentin2105** [comment]
@leblancd Thanks for your IPv6 work on k8s !!
Any idea in which release Kube-proxy will start to handle IP6table for Services ? and specially Service/ExternalIPs ?

**Comment by dims** [comment]
cc @sadasu

**Comment by leblancd** [comment]
I have created a [Kubernetes IPv6 deployment guide](https://github.com/leblancd/kube-v6) based on a forked release containing several outstanding PR's. The guide will be transitioned into upstream Kubernetes documentation for the 1.9 release. Feel free to use the guide in the interim for test/dev efforts. I would appreciate any feedback.

**Comment by leblancd** [comment]
@valentin2105 : Services/externalIPs should work with required cherry picks. See the [Kubernetes IPv6 deployment guide](https://github.com/leblancd/kube-v6) which refers to [Kubernetes IPv6 Version v1.9.0-alpha.0.ipv6.0](https://github.com/leblancd/kubernetes/releases/tag/v1.9.0-alpha.0.ipv6.0). I would appreciate any feedback.

**Comment by lachie83** [comment]
@leblancd thank you very much for putting this together!

**Comment by leblancd** [comment]
@lachie83 : YW on behalf of the IPv6 working group: @danehans, @pmichali, @rpothier, @aanm, and the sig-network team.

**Comment by valentin2105** [comment]
I wrote this post if it can help people to Deploy IPv6 in Kubernetes // https://opsnotice.xyz/kubernetes-ipv6-only/

**Comment by burtonator** [comment]
What's needed to get this implemented.  We would love to see IPv6 at Scalefastr as that works really well for our use case.  

IPv6 is awesome when you have a whole /64 and plenty of IPs to work with on the host machine.

**Comment by danehans** [comment]
@burtonator IPv6 will be added as an alpha feature in the 1.9 release. You can use kubeadm to deploy an IPv6 Kubernetes cluster by specifying an IPv6 address for `--apiserver-advertise-address` and using brackets around the IPv6 master address for `kubeadm join --token <token> [<master-ip>]:<master-port>`. The above information and other specifics will be part of the 1.9 release documentation. Prior to 1.9, @leblancd create the [kube-v6 project](https://github.com/leblancd/kube-v6/blob/master/README.md) to test Kubernetes with IPv6.

**Comment by valentin2105** [comment]
I think ClusterIP is an important point about IPv6 : 

https://github.com/kubernetes/kubernetes/issues/57069

**Comment by leblancd** [comment]
@burtonator What CNI network plugin do you use?

**Comment by leblancd** [comment]
This issue has been used to follow support for IPv6-only clusters (Alpha support in Kubernetes 1.9)
For dual-stack support, new issues have been filed and a design document has been proposed:
    https://github.com/kubernetes/kubernetes/issues/62822
    https://github.com/kubernetes/features/issues/563
    https://github.com/kubernetes/community/pull/2254
